### PR TITLE
Provider client context manager

### PR DESCRIPTION
**Pull request by fbalak**
Simplify context usage for Provider client setups.

### CODE REVIEW

**Comment by fbalak** [comment]
Verified: https://ocs4-jenkins-csb-odf-qe.apps.ocp-c1.prod.psi.redhat.com/job/qe-odf-multicluster/2630/

**Review by DanielOsypenko — APPROVED** [review]

**Review by jilju — DISMISSED** [review]

File: ocs_ci/framework/__init__.py [hunk 1]
```diff
@@ -479,6 +481,36 @@ def __init__(self):
             primary_index = primary_config.MULTICLUSTER.get("multicluster_index")
             super().__init__(primary_index)
 
+    class RunWithProviderConfigContext(RunWithConfigContext):
+        """
+        Context manager that makes sure that a given code block is executed on Provider.
```
Comment (petr-balogh) [inline ocs_ci/framework/__init__.py:486]: I would mention in docstring that if no provider cluster found, the code will run on current index.
And not sure we should change name of class to something like:
`RunWithProviderOrCurrnetConfigContext`

BTW This logic makes sense only if we would like to have this code using this context manager to work in provider/consumer as well as other type of executions, correct?

File: ocs_ci/framework/__init__.py [hunk 2]
```diff
@@ -479,6 +481,36 @@ def __init__(self):
...
+            except ClusterNotFoundException:
+                # if no provider is available then set the switch to current index so that
+                # no switch happens and code runs on current cluster
+                logger.DEBUG("No provider was found - using current cluster")
+                switch_index = config.cur_index
+            super().__init__(switch_index)
+
+    class RunWithFirstConsumerConfigContext(RunWithConfigContext):
+        """
+        Context manager that makes sure that a given code block is executed on First consumer.
```
Comment (petr-balogh) [inline ocs_ci/framework/__init__.py:501]: Same as above comment applies here.

File: ocs_ci/ocs/resources/storageconsumer.py [hunk 3]
```diff
@@ -86,53 +87,36 @@ def stop_heartbeat(self):
         """
         Suspend status reporter cron job.
         """
-        self._switch_consumer_cluster()
-        patch_param = '{"spec": {"suspend": true}}'
-        self.heartbeat_cronjob.ocp.patch(
-            resource_name=self.heartbeat_cronjob.name, params=patch_param
-        )
-        self._switch_provider_cluster()
+        with config.RunWithConfigContext(self.consumer_context):
```
Comment (petr-balogh) [inline ocs_ci/ocs/resources/storageconsumer.py:90]: Do we hold some specific consumer context in `self.consumer_context`?
Or why not using new context manager `RunWithFirstConsumerConfigContext` here and other occurrences below?

Comment (fbalak) [reply ocs_ci/ocs/resources/storageconsumer.py:90]: It uses `consumer_context` defined for the instance constructor of the `StorageConsumer` class. It doesn't have to be the first consumer in all cases.

File: ocs_ci/framework/__init__.py (hunk 1 above)
Comment (fbalak) [reply ocs_ci/framework/__init__.py:486]: > I would mention in docstring that if no provider cluster found, the code will run on current index.

This makes sense.

> And not sure we should change name of class to something like:
> RunWithProviderOrCurrnetConfigContext

How about `RunWithProviderConfigContextIfAvailable`?

> BTW This logic makes sense only if we would like to have this code using this context manager to work in provider/consumer as well as other type of executions, correct?

Yes, this way people can add it to the code without a worry that it breaks other test cases. It allows to mark chunks of code with a context for provider/consumer in generic function e.g. in `ocs_ci/ocs/resources/*` or `tests/conftest.py`

**Review by petr-balogh — APPROVED** [review]

**Comment by petr-balogh** [comment]
/cherry-pick release-4.16

I think we need to cherry pick to 4.16 at least, right @fbalak ?

**Review by DanielOsypenko — APPROVED** [review]

**Pull request merged**